binary search tree code added

sumitrai opened the pull request:
@HackerHorse Binary search tree beta implementation after reading part of tutorial from:
http://cslibrary.stanford.edu/110/BinaryTrees.html (Nick P. guy who also has awesome google python classes on youtube). Some functions still might need to be added, this is just first version.
- main.c: now test binary tree library
- include/binary_tree.h: new file, binary tree header
- lib/binary_tree.c: new file, binary search tree implementation
  (bt_new_node): create a new binary tree node on heap
  (bt_insert): inser a new node into binary tree
  (bt_lookup): lookup a node into binary tree

Note: test with
make main
./main

Commit:
binary search tree code added

* main.c: now test binary tree library

* include/binary_tree.h: new file, binary tree header

* lib/binary_tree.c: new file, binary search tree implementation
(bt_new_node): create a new binary tree node on heap
(bt_insert): inser a new node into binary tree
(bt_lookup): lookup a node into binary tree

Note: test with
make main
./main

prashinde commented on main.c:
```diff
@@ -1,21 +1,23 @@
-    add_node_pos(&headp, create_node(200), 200);
-    print_all_slist(headp);
+    // Note: Previous Linked-List feedback applied here - root is now on heap
+    struct bt_node *root = bt_new_node(10);
```
What if bt_new_node() failed to allocate memory. Handle failure here.
Read about "errno".

sumitrai replied:
Thanks for pointing that out, latest push now does assert checks and uses strerror() to print reason for memory allocation failure.

prashinde commented on lib/binary_tree.c:
```diff
@@ -0,0 +1,42 @@
+void bt_insert(struct bt_node **node, int val)
+{
+    if (*node == NULL)
+        *node = bt_new_node(val);
```
Dangerous call! Let the caller know possible failure in bt_new_node() or just ASSERT().

sumitrai replied:
main now checks if root is NULL and prints error via strerror(errno), returns 1.

sssharma10 commented on lib/binary_tree.c:
```diff
@@ -0,0 +1,42 @@
+
+void bt_insert(struct bt_node **node, int val)
+{
+    if (*node == NULL)
```
What is node is NULL, there is no check for that.

sumitrai replied:
added assert check for this.

Commit:
pull request comments resolved

* main.c (main): print error to stderr and return 1 if root node allocation fails.

* lib/binary_tree.c (bt_insert): added assert check for node argument and *node
after it's allocated.

Commit:
prints reason for memory allocation failsure via errnor

* main.c (main): print error to stderr and return 1 if root node allocation fails.
(main): now prints reason for memory allocation failure via strerror().

Commit:
added errno intrepretation to lib/

* lib/binary_tree.c (bt_insert): added assert check for node argument and *node
after it's allocated.
(bt_new_node): now prints reason for memory allocation failrue vai strerror().

prashinde commented on lib/binary_tree.c:
```diff
@@ -8,6 +12,9 @@ struct bt_node *bt_new_node(int val)
     if (bt_node_ptr) {
         bt_node_ptr->data = val;
         bt_node_ptr->left = bt_node_ptr->right = NULL;
+    } else {
```
(style) Better to return an error in "if" clause than "else".

struct bt_node *bt_node_ptr = malloc(sizeof(struct bt_node));
      if (bt_node_ptr == NULL) {
          fprintf("ERROR");
          return NULL; //or goto out to keep single exit point
      }
      //contines here

sumitrai replied:
now, all functions return on failure and flow continues otherwise.
Changed insert and new_node function prototypes to return 1 in case of error and every library function now checks if underlying operation was success.

prashinde commented on lib/binary_tree.c:
```diff
@@ -8,6 +12,9 @@ struct bt_node *bt_new_node(int val)
     if (bt_node_ptr) {
         bt_node_ptr->data = val;
         bt_node_ptr->left = bt_node_ptr->right = NULL;
+    } else {
```
It's just a style which kernel follows. You can chose to differ.

Commit:
added more error handling code as per Pratik's comments

* main.c (main): added more error handling code and updated code as per
new function prototypes.

* lib/binary_tree.c (bt_new_node): prototype changed and added error handling code.
now returns 1 on failure, 0 on success.
(bst_insert, bst_lookup): renamed from bt_*, and See (bt_new_node)

Commit:
Merge branch 'master' of github.com:sumitrai/ds_and_algo
